### PR TITLE
[MI-2165]: Implemented Admin management flow for subscriptions.

### PR DESCRIPTION
**Pull request by avas27JTG**
Implemented Admin management flow for subscriptions.

image

image

image

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -255,6 +255,13 @@ func (p *Plugin) handleCreateSubscription(w http.ResponseWriter, r *http.Request
 		return
 	}
 
+	user, userErr := p.API.GetUser(mattermostUserID)
+	if userErr != nil {
+		p.API.LogError(constants.GetUserError, "Error", userErr.Error())
+		http.Error(w, fmt.Sprintf("%s. Error: %s", constants.GetUserError, userErr.Error()), userErr.StatusCode)
```
Comment (ayusht2810) [inline server/plugin/api.go:261]: we can use `p.handleError` here

File: server/store/subscriptions.go [hunk 2]
```diff
@@ -95,9 +97,18 @@ func (s *Store) GetAllSubscriptions(userID string) ([]*serializers.SubscriptionD
...
+	if userID == "" {
+		for mmUserID := range subscriptions.ByMattermostUserID {
+			for _, subscription := range subscriptions.ByMattermostUserID[mmUserID] {
+				subscription := subscription // we need to do this to prevent implicit memory aliasing in for loop
+				subscriptionList = append(subscriptionList, &subscription)
+			}
+		}
+	} else {
+		for _, subscription := range subscriptions.ByMattermostUserID[userID] {
+			subscription := subscription // we need to do this to prevent implicit memory aliasing in for loop
```
Comment (ayusht2810) [inline server/store/subscriptions.go:109]:
```suggestion
			subscription := subscription
```

Comment (ayusht2810) [reply server/store/subscriptions.go:109]: we should add the same comment once per file

File: webapp/src/containers/Rhs/projectDetails/header.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
+import React, {useState} from 'react';
+import {useDispatch} from 'react-redux';
+
+import plugin_constants from 'plugin_constants';
+
+import BackButton from 'components/buttons/backButton';
+import PrimaryButton from 'components/buttons/primaryButton';
+import ConfirmationModal from 'components/modal/confirmationModal';
+import ToggleSwitch from 'components/toggleSwitch';
+import Dropdown from 'components/dropdown';
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectDetails/header.tsx:10]: is the import order correct here?

File: webapp/src/plugin_constants/messages.ts [hunk 4]
```diff
@@ -7,4 +7,10 @@ export const error = {
 
     // Subscription
     subscriptionAlreadyExists: 'Requested subscription already exists',
+    accessDenied: 'Access Denied',
+    subscriptionNotFound: 'Requested subscription does not exists',
+    adminAccessError: 'Looks like you do not have access to add/delete subscription for this project. Please make sure you are a project or team administrator for this project',
```
Comment (ayusht2810) [inline webapp/src/plugin_constants/messages.ts:12]:
```suggestion
    adminAccessError: 'Looks like you do not have access to add/delete a subscription for this project. Please make sure you are a project or team administrator for this project.',
```

File: webapp/src/plugin_constants/messages.ts [hunk 5]
```diff
@@ -7,4 +7,10 @@ export const error = {
...
+
+    // Link
+    notAccessibleError: 'Looks like this project/organization does not exists or you do not have permissions to access it',
```
Comment (ayusht2810) [inline webapp/src/plugin_constants/messages.ts:15]:
```suggestion
    notAccessibleError: 'Looks like this project/organization does not exist or you do not have permissions to access it.',
```

File: webapp/src/containers/Rhs/projectDetails/header.tsx (hunk 3 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/header.tsx:10]: there is no hard rule on import order for same type like we can import all `components` together and their order matters but inside that I don't think any rule is set

File: webapp/src/plugin_constants/messages.ts (hunk 5 above)
Comment (avas27JTG) [reply webapp/src/plugin_constants/messages.ts]: we haven't put any full stop after any message on webapp
also, its a standard practise across many products that they avoid `.` for message

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -802,7 +805,8 @@ func TestHandleDeleteSubscriptions(t *testing.T) {
 				"organization": "mockOrganization",
 				"project": "mockProject",
 				"eventType": "mockEventType",
-				"channelID": "mockChannelID"
+				"channelID": "mockChannelID",
+				"mmUserID":"mockMattermostUserID"
```
Comment (ayusht2810) [inline server/plugin/api_test.go:809]:
```suggestion
				"mmUserID": "mockMattermostUserID"
```

**Review by ayusht2810 — APPROVED** [review]

File: webapp/src/containers/Rhs/projectDetails/header.tsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                    />
+                </div>
+                <div className='d-flex align-item-center'>
+                    <ToggleSwitch
+                        active={showAllSubscriptions}
+                        onChange={(active) => {
+                            handlePagination(true, false);
+                            setShowAllSubscriptions(active);
+                        }}
+                        label={'Show All Subscriptions'}
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectDetails/header.tsx:88]: Remember to change this label according to the latest changes suggested by Mike.

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 8]
```diff
@@ -37,9 +36,10 @@ const ProjectDetails = memo((projectDetails: ProjectDetails) => {
     });
     const [subscriptionList, setSubscriptionList] = useState<SubscriptionDetails[]>([]);
     const [showAllSubscriptions, setShowAllSubscriptions] = useState(false);
-    const [showProjectConfirmationModal, setShowProjectConfirmationModal] = useState(false);
+    const [filter, setFilter] = useState(plugin_constants.common.SubscriptionFilterCreatedBy.me);
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:39]: Can you extract the required variables from this at the top?
For eg - 
```
const {SubscriptionFilterCreatedBy} = plugin_constants.common;
```

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 9]
```diff
@@ -139,8 +118,24 @@ const ProjectDetails = memo((projectDetails: ProjectDetails) => {
             handlePagination(true);
             setShowSubscriptionConfirmationModal(false);
         },
+        handleError: (error) => {
+            const errorMessage = utils.getErrorMessage(true, 'ConfirmationModal', error);
+            if (errorMessage === plugin_constants.messages.error.subscriptionNotFound) {
+                handlePagination(true);
+                setShowSubscriptionConfirmationModal(false);
+                return;
+            }
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:127]:
```suggestion
            }
            
```

File: webapp/src/plugin_constants/messages.ts [hunk 10]
```diff
@@ -7,4 +7,10 @@ export const error = {
 
     // Subscription
     subscriptionAlreadyExists: 'Requested subscription already exists',
+    accessDenied: 'Access Denied',
+    subscriptionNotFound: 'Requested subscription does not exists',
```
Comment (manojmalik20) [inline webapp/src/plugin_constants/messages.ts:11]:
```suggestion
    subscriptionNotFound: 'Requested subscription does not exist',
```

File: webapp/src/types/common/index.d.ts [hunk 11]
```diff
@@ -64,9 +65,16 @@ type SubscriptionDetails = {
...
+    createdBy: string,
+    username: string,
 }
 
 type WebsocketEventParams = {
     event: string,
     data: Record<string, string>,
 }
+
+type ConfirmationModalErrorPanel = {
```
Comment (manojmalik20) [inline webapp/src/types/common/index.d.ts:77]:
```suggestion
type ConfirmationModalErrorPanelProps = {
```

File: server/plugin/api.go [hunk 12]
```diff
@@ -275,10 +284,19 @@ func (p *Plugin) handleCreateSubscription(w http.ResponseWriter, r *http.Request
...
+
+	var subscriptionList []*serializers.SubscriptionDetails
+	var subscriptionErr error
+	createdBy := r.URL.Query().Get(constants.QueryParamCreatedBy)
+	switch {
+	case createdBy == constants.FilterCreatedByMe || createdBy == "":
+		subscriptionList, subscriptionErr = p.Store.GetAllSubscriptions(mattermostUserID)
+	case createdBy == constants.FilterCreatedByAnyone:
+		subscriptionList, subscriptionErr = p.Store.GetAllSubscriptions("")
+	}
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply server/plugin/api.go]: FYI: combining two cases this way in GO didn't work
```
case constants.FilterCreatedByMe:
case "":
```

did this way
```
case constants.FilterCreatedByMe, "":
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**